src/ILCompiler.Compiler/src/Compiler/SingleFileCompilationModuleGroup.cs

```diff
@@ -27,7 +27,11 @@ public override void AddCompilationRoots(IRootingServiceProvider rootProvider)
         {
             base.AddCompilationRoots(rootProvider);
 
-            AddCompilationRootsForExports((EcmaModule)_typeSystemContext.SystemModule, rootProvider);
+            foreach (var inputFile in _typeSystemContext.ReferenceFilePaths)
+            {
```

Comment:
The single file compilation is meant to mimic simplicity and efficiency of the classic compilers. You can do `cl hello.cpp` and it will quickly produce self-contained .exe. Underneath, the compiler has references to a ton of standard header files and the linker command line has references to number of .libs - so that you do not need to worry about adding just the right ones. You won't get any of it unless it is referenced from your .cpp file.

The simple things should be simple, the hard things should be possible. If you have multiple libraries with named NativeCallable exports, I think it is fine that you need to do a bit extra work for it. Also, I would not mind if we introduce more fine grained controls for it (e.g. scan named exports for given module, add full reflection for given module, etc.). But I do not think we should be adding roots for everything in the reference list by default.